ingress/controllers/nginx/controller.go

```diff
@@ -217,6 +248,126 @@ func (lbc *loadBalancerController) getUDPConfigMap(ns, name string) (*api.Config
 	return lbc.client.ConfigMaps(ns).Get(name)
 }
 
+func (lbc *loadBalancerController) updateEpNamedPorts(key string) {
+	if !lbc.controllersInSync() {
+		lbc.svcEpQueue.requeue(key, fmt.Errorf("deferring sync till endpoints controller has synced"))
+		return
+	}
+
+	glog.V(4).Infof("checking if service %v uses named ports to update annotation %v", key, namedPortAnnotation)
+
+	ingObj, ingExists, err := lbc.ingLister.Store.GetByKey(key)
+	if err != nil {
+		glog.Warningf("error getting service %v: %v", key, err)
+		return
+	}
+
+	if !ingExists {
+		glog.Warningf("service %v not found", key)
+		return
+	}
+
+	ing := ingObj.(*extensions.Ingress)
+	for _, rule := range ing.Spec.Rules {
+		if rule.IngressRuleValue.HTTP == nil {
+			continue
+		}
+
+		for _, path := range rule.HTTP.Paths {
+			svcKey := fmt.Sprintf("%v/%v", ing.GetNamespace(), path.Backend.ServiceName)
+			svcObj, svcExists, err := lbc.svcLister.Store.GetByKey(svcKey)
+			if err != nil {
+				glog.Infof("error getting service %v from the cache: %v", svcKey, err)
+				continue
+			}
+
+			if !svcExists {
+				glog.Warningf("service %v does no exists", svcKey)
+				continue
+			}
+
+			svc := svcObj.(*api.Service)
+			if svc.Spec.Selector == nil {
+				return
+			}
+
+			err = lbc.checkSvcForUpdate(svc)
+			if err != nil {
+				lbc.svcEpQueue.requeue(key, err)
+				return
+			}
+		}
+	}
+}
+
+// checkSvcForUpdate verifies if one of the running pods for a service contains
+// named port. If the annotation in the service does not exists or is not equals
+// to the port mapping obtained from the pod the service must be updated to reflect
+// the current state
+func (lbc *loadBalancerController) checkSvcForUpdate(svc *api.Service) error {
+	// get the pods associated with the service
+	pods, err := lbc.client.Pods(svc.Namespace).List(api.ListOptions{
+		LabelSelector: labels.Set(svc.Spec.Selector).AsSelector(),
+	})
+	if err != nil {
+		return fmt.Errorf("error searching service pods %v/%v: %v", svc.Namespace, svc.Name, err)
+	}
+
+	if len(pods.Items) == 0 {
+		return nil
+	}
+
+	namedPorts := map[string]string{}
+	// we need to check only one pod searching for named ports
+	pod := &pods.Items[0]
+	glog.V(4).Infof("checking pod %v/%v for named port information", pod.Namespace, pod.Name)
+	for i := range svc.Spec.Ports {
+		servicePort := &svc.Spec.Ports[i]
+
+		_, err := strconv.Atoi(servicePort.TargetPort.StrVal)
```

Comment:
can you bail early (before pod list) if none of these ports is a string?

If we can't selectively list only when there's a named port, I'd suggest setting up a pod watch, eg: 
```go
	podStore.Store, podController = framework.NewInformer(
 		&cache.ListWatch{
 			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
 				return e.client.Core().Pods(api.NamespaceAll).List(options)
 			},
 			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
 				return e.client.Core().Pods(api.NamespaceAll).Watch(options)
 			},
 		},
 		&api.Pod{},
 		resyncPeriod(),
 		framework.ResourceEventHandlerFuncs{
 			AddFunc:    addPod,
 			UpdateFunc: updatePod,
 			DeleteFunc: deletePod,
 		},
 	)
```
And then you have a local cache you can list for pods:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L542